### ISSUE TITLE
Generic monad algorithm

### ISSUE THREAD

**Issue by denghz** [dbrattli/Expression#100]
Is it possible to define a function that can handle multiple types of effects depends on the object that yields the effect?
Like in Haskell
``` 
join :: Monad m => m (m a) -> m a
join xss = 
    do xs <- xss
       xs                            
```
which would work for all monads including, `list`, `maybe`, `io`, `stm`, etc...
I can help about it if needed.

**Comment by dbrattli** [comment]
No, I don't think it's possible e.g an iterable join is very different from an [observable join](https://github.com/dbrattli/aioreactive/blob/master/aioreactive/combine.py#L60) so each world will need a specific implementation. It will not be possible for the specific implementation to be registered as an auto-dispatch / overload. But if you have a suggestion then please feel free to explain what idea you have in mind.

**Comment by dbrattli** [comment]
Closing as answered. Effect-specific operations have materially different semantics and cannot be safely expressed as a single generic monad algorithm with Python dispatch. No alternative design proposal followed the prior explanation.
